kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,771 @@
+                             (hit_level and hit_level < HIT_L3))
+    end
+
+    stats_count(self.stats, name, HIT_LEVEL_TO_NAME[hit_level])
+  end
+
+  -- dereference CNAME
+  if qtype ~= TYPE_CNAME and answers and answers[1].type == TYPE_CNAME then
+    stats_count(self.stats, name, "cname")
+    return resolve_all(self, answers[1].cname, qtype, cache_only, tries, resolved_names)
+  end
+
+  return answers, err, tries
+end
+
+
+function _M:resolve(name, qtype, cache_only, tries)
+  return resolve_all(self, name, qtype, cache_only, tries, {})
+end
+
+
+function _M:resolve_address(name, port, cache_only, tries, resolved_names)
+  resolved_names = resolved_names or {}
+
+  local answers, err, tries = resolve_all(self, name, nil, cache_only, tries, resolved_names)
+  if not answers then
+    return nil, err, tries
+  end
+
+  -- non-nil answers and return_random
+  if answers[1].type == TYPE_SRV then
+    local answer = get_weighted_round_robin_answers(answers)
+    port = (answer.port ~= 0 and answer.port) or port
+    return self:resolve_address(answer.target, port, cache_only, tries, resolved_names)
+  end
+
+  return get_next_round_robin_answers(answers).address, port, tries
+end
+
+
+-- compatible with original DNS client library
+-- These APIs will be deprecated if fully replacing the original one.
+local dns_client
+
+function _M.init(opts)
+  opts = opts or {}
+  opts.valid_ttl = opts.validTtl
+  opts.error_ttl = opts.badTtl
+  opts.stale_ttl = opts.staleTtl
+  opts.cache_size = opts.cacheSize
+
+  local client, err = _M.new(opts)
+  if not client then
+    return nil, err
+  end
+
+  dns_client = client
+  return true
+end
+
+
+-- New and old libraries have the same function name.
+_M._resolve = _M.resolve
+
+function _M.resolve(name, r_opts, cache_only, tries)
+  return dns_client:_resolve(name, r_opts and r_opts.qtype, cache_only, tries)
+end
+
+
+function _M.toip(name, port, cache_only, tries)
+  return dns_client:resolve_address(name, port, cache_only, tries)
+end
+
+
+-- for example, "example.com:33" -> "example.com:SRV"
+local function format_key(key)
+  local qname, qtype = key:match("([^:]+):(%d+)")  -- match "(qname):(qtype)"
+  return qtype and qname .. ":" .. (TYPE_TO_NAME[tonumber(qtype)] or qtype)
+               or  key
+end
+
+
+function _M.stats()
+  local stats = {}
+  for k, v in pairs(dns_client.stats) do
+    stats[format_key(k)] = v
+  end
+  return stats
+end
+
+
+-- For testing
+
+if package.loaded.busted then
+  function _M.getobj()
+    return dns_client
+  end
+
+  function _M.getcache()
+    return {
+      set = function(self, k, v, ttl)
+        self.cache:set(k, {ttl = ttl or 0}, v)
+      end,
+
+      delete = function(self, k)
+        self.cache:delete(k)
+      end,
+
+      cache = dns_client.cache,
+    }
+  end
+
+  function _M:insert_last_type(name, qtype)
+    insert_last_type(self.cache, name, qtype)
+  end
+
+  function _M:get_last_type(name)
+    return get_last_type(self.cache, name)
+  end
```

Comment:
```suggestion
  function _M:_insert_last_type(name, qtype)  -- export as different name!
    insert_last_type(self.cache, name, qtype)
  end

  function _M:_get_last_type(name)            -- export as different name!
    return get_last_type(self.cache, name)
  end
```
**Always change the name** if you're exporting a local for testing!!

If your code accidentally calls `client.get_last_type(client, ...)` instead of `get_last_type(client, ...)` then it will pass the tests, but fail once deployed because that method no longer exists. Very tricky!

see also https://github.com/Kong/kong/blob/refactor/dns_client/kong/resty/dns/client.lua#L1600

Reply:
cool, resolved!